Services/CustomDeviceModels.cs

```diff
@@ -0,0 +1,222 @@
+                    () => new { e.Message, Exception = e });
+                throw new ExternalDependencyException("Unable to parse device models loaded from storage", e);
+            }
+        }
+
+        /// <summary>
+        /// Get a custom device model.
+        /// </summary>
+        public async Task<DeviceModel> GetAsync(string id)
+        {
+            if (string.IsNullOrEmpty(id))
+            {
+                this.log.Error("Device model id cannot be empty!", () => { });
+                throw new InvalidInputException("Device model id cannot be empty! ");
+            }
+
+            var item = await this.storage.GetAsync(STORAGE_COLLECTION, id);
+            var deviceModel = JsonConvert.DeserializeObject<DeviceModel>(item.Data);
+            deviceModel.ETag = item.ETag;
+            return deviceModel;
+        }
+
+        /// <summary>
+        /// Create a custom device model.
+        /// </summary>
+        public async Task<DeviceModel> InsertAsync(DeviceModel deviceModel, bool generateId = true)
+        {
+            deviceModel.Created = DateTimeOffset.UtcNow;
+            deviceModel.Modified = deviceModel.Created;
+
+            if (generateId)
+            {
+                deviceModel.Id = Guid.NewGuid().ToString();
+            }
+
+            this.log.Debug("Creating a custom device model.", () => new { deviceModel });
+
+            try
+            {
+                // Note: using UpdateAsync because the service generates the ID
+                var result = await this.storage.UpdateAsync(
+                    STORAGE_COLLECTION,
+                    deviceModel.Id,
+                    JsonConvert.SerializeObject(deviceModel),
+                    null);
+
+                deviceModel.ETag = result.ETag;
+            }
+            catch (Exception e)
+            {
+                this.log.Error("Failed to insert new device model into storage",
+                    () => new { deviceModel, generateId, e });
+                throw new ExternalDependencyException(
+                    "Failed to insert new device model into storage", e);
+            }
+
+            return deviceModel;
+        }
+
+        /// <summary>
+        /// Create or replace a custom device model.
+        /// </summary>
+        public async Task<DeviceModel> UpsertAsync(DeviceModel deviceModel)
+        {
+            var id = deviceModel.Id;
+            var eTag = deviceModel.ETag;
+
+            try
+            {
+                var item = await this.GetAsync(id);
+
+                if (item.ETag == eTag)
+                {
+                    // Replace a custom device model
+                    deviceModel.Created = item.Created;
+                    deviceModel.Modified = DateTimeOffset.UtcNow;
+
+                    this.log.Debug("Modifying a custom device model via PUT.", () => new { deviceModel });
+
+                    var result = await this.storage.UpdateAsync(
```

Comment:
similarly to InsertAsync, we need a test to verify the data written to storage is correct and under the right id

Reply:
Done.

Reply:
I can't see this test, could you point me to it? is the PR up to date?